### PR TITLE
[Seq] Computational register op and lowering to SV

### PR DESCRIPTION
**Pull request by teqdruid**
Model of the `compreg` op outlined in the Seq dialect rationale dialect. This is just a starting place and I do expect it to change and/or have siblings for different flavors. Lowers to `sv.alwaysff` with posedge clock and sync reset -- for now.

Also adds a few mechanical things required to make the Seq dialect actually work which I neglected in the PR which added the dialect.

### CODE REVIEW

File: lib/Dialect/SV/SVOps.cpp [hunk 1]
```diff
@@ -155,7 +155,7 @@ void RegOp::build(OpBuilder &odsBuilder, OperationState &odsState,
 void RegOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
   // If the wire has an optional 'name' attribute, use it.
   auto nameAttr = (*this)->getAttrOfType<StringAttr>("name");
-  if (!nameAttr.getValue().empty())
+  if (nameAttr && !nameAttr.getValue().empty())
```
Comment (teqdruid) [inline lib/Dialect/SV/SVOps.cpp:158]: This is an unrelated small fix for a bug I encountered during development.

Comment (lattner) [reply lib/Dialect/SV/SVOps.cpp:158]: This is (now) required in the sv.reg op, so please drop this:

` let arguments = (ins StrAttr:$name);`

File: lib/Dialect/Seq/SeqOps.cpp [hunk 2]
```diff
@@ -12,11 +12,62 @@
...
+  }
+
+  if (parser.parseOptionalAttrDict(result.attributes))
+    return failure();
+
+  if (parser.parseColon())
+    return failure();
+
+  Type ty;
+  if (parser.parseType(ty))
```
Comment (lattner) [inline lib/Dialect/Seq/SeqOps.cpp:50]: Please merge all these together to reduce lines of code and make it more clear:

if (parseoptionalattrdict() ||
    parseColon||
   parseType )
  return failure

File: lib/Dialect/Seq/SeqPasses.cpp [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    auto svReg = rewriter.create<sv::RegOp>(loc, reg.getResult().getType());
+    DictionaryAttr regAttrs = reg->getAttrDictionary();
+    if (!regAttrs.empty())
+      svReg->setAttrs(regAttrs);
+    if (!svReg->hasAttrOfType<StringAttr>("name"))
+      // sv.reg requires a name attribute.
+      svReg->setAttr("name", rewriter.getStringAttr(""));
+    auto regVal = rewriter.create<sv::ReadInOutOp>(loc, svReg);
+
+    if (reg.reset() && reg.resetValue())
```
Comment (lattner) [inline lib/Dialect/Seq/SeqPasses.cpp:54]: nit but please put braces around these conditions since they are complicated.

**Review by lattner — APPROVED** [review]

File: lib/Dialect/SV/SVOps.cpp (hunk 1 above)
Comment (teqdruid) [reply lib/Dialect/SV/SVOps.cpp:158]: You mean from the RegOp ODS? If it's required, why should I delete it?

This was simply to avoid a crash with "--debug" while debugging my lowering pass.

Comment (teqdruid) [reply lib/Dialect/SV/SVOps.cpp:158]: I think there are certain functions which should never crash including on invalid IR. It should at least be an assert. It wasn't clear that this was the issue.

Comment (teqdruid) [reply lib/Dialect/SV/SVOps.cpp:158]: Sorry, you meant drop the change, not delete the `let arguments = (ins StrAttr:$name);` line from the RegOp ODS, yes?

**Comment by teqdruid** [comment]
@lattner Thanks for the review!
